File: 31-03-2025-Maximize_Partitions_in_a_String.java
class Solution {
    public int maxPartitions(String s) {
        int n = s.length();
        int[] lastIndex = new int[26];
        List<Integer> ans = new ArrayList<>();

        for (int i = 0; i < n; ++i) {
            lastIndex[s.charAt(i) - 'a'] = i;
        }

        int farthest = 0, left = 0;
        for (int i = 0; i < n; ++i) {
            farthest = Math.max(farthest, lastIndex[s.charAt(i) - 'a']);
            if (i == farthest) {
                ans.add(farthest - left + 1);
                left = i + 1;
            }
        }
        return ans.size();
    }
}

File: 31-03-2025-Maximize_Partitions_in_a_String.py
class Solution:

     def maxPartitions(self, s: str) -> int:
        n = len(s)
        last_index = {ch: i for i, ch in enumerate(s)}
        ans = []
        
        farthest = 0
        left = 0
        for i, ch in enumerate(s):
            farthest = max(farthest, last_index[ch])
            if i == farthest:
                ans.append(farthest - left + 1)
                left = i + 1
        
        return len(ans)
